### PR TITLE
Make `#[public]` ffi functions private and unsafe

### PR DESCRIPTION
**Pull request by richardpringle**
Make it harder to call the `extern` functions written by the `#[public]` macro

### CODE REVIEW

File: x/programs/rust/sdk_macros/Cargo.toml
Comment (richardpringle) [inline x/programs/rust/sdk_macros/Cargo.toml:1]: Needed for the `ui` test

File: x/programs/rust/sdk_macros/src/lib.rs
Comment (richardpringle) [inline x/programs/rust/sdk_macros/src/lib.rs:1]: Swap the user-specified type for the fully qualified type. If the type is actually a user-defined type with the same name, we now get the proper error.

File: x/programs/rust/sdk_macros/src/lib.rs [hunk 1]
```diff
@@ -64,11 +68,11 @@ pub fn public(_: TokenStream, item: TokenStream) -> TokenStream {
         }
     };
 
-    let param_idents = input_args
-        .iter()
-        .enumerate()
-        .skip(1)
-        .map(|(i, fn_arg)| match fn_arg {
+    let (input, context_type) = (input, context_type);
```
Comment (iFrostizz) [inline x/programs/rust/sdk_macros/src/lib.rs:71]: nit: this is a noop

**Review by iFrostizz — APPROVED** [review]
nice!

File: x/programs/rust/sdk_macros/src/lib.rs (hunk 1 above)
Comment (richardpringle) [reply x/programs/rust/sdk_macros/src/lib.rs:71]: Not quite, it's a move from mutable to immutable... but it's not obvious, so I'll make a change here

**Review by iFrostizz — APPROVED** [review]

**Pull request merged**